compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1521,7 +1521,7 @@ public boolean validateKeySpecifier(List<String> fieldNameList, BType constraint
                 return true;
             }
 
-            if (!Symbols.isFlagOn(field.symbol.flags, Flags.REQUIRED)) {
+            if (Symbols.isFlagOn(field.symbol.flags, Flags.OPTIONAL)) {
```

Comment:
`table<record {readonly int id = 1; string name;}> key (id) tb = table [];`
When there is a default value, the flag set does not contain `FILED` and `REQUIRED`, but those flags are available when there is no default value.
Isn't that what we need to fix?
@gimantha , @pcnfernando WDYT?

Reply:
Yes with this change, even optional keys will be accepted for key specifiers. If/when we decide to support default values with the table constructor, the check should not be for `Optional` but for `isDefaultable`

Reply:
FIELD should be available.

```ballerina
type Foo record {
   int a;
   int b?;
   int c = 1;
};
```

But IIRC, REQUIRED and OPTIONAL are set as follows:
- `a` - REQUIRED
- `b` - OPTIONAL
- `c` - neither `REQUIRED` nor `OPTIONAL`
which is how we distinguish between the three kinds of fields.

Reply:
Then we can change the error message like key specifier cannot be optional.

Reply:
Created a issue for this also. https://github.com/ballerina-platform/ballerina-lang/issues/37624